### PR TITLE
CLOUDP-249804: Add Not found to search state machine

### PR DESCRIPTION
**Pull request by josvazg**
Extending the [state machine to handle the not found error case as well explicitly](https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0A%20%20%20%20pending%20-%3E%20pending%20%5Blabel%3D%22terminate%20(failed)%22%5D%0A%20%20%20%20%0A%20%20%20%20pending%20-%3E%20in_progress%5Blabel%3D%22busy%22%5D%20%2F%2F%20Atlas%20update%20still%20ongoing%0A%20%20%20%20%0A%20%20%20%20pending%20-%3E%20in_progress%5Blabel%3D%22update%22%5D%20%2F%2F%20Atlas%20state%20does%20not%20match%20the%20spec%0A%20%20%20%20%0A%20%20%20%20pending%20-%3E%20idle%20%2F%2F%20Atlas%20state%20matches%20spec%0A%20%20%20%20%0A%20%20%20%20pending%20-%3E%20deleted%5Blabel%3D%22delete%22%5D%20%2F%2F%20In%20Atlas%20but%20not%20in%20spec%0A%20%20%20%20%0A%20%20%20%20pending%20-%3E%20deleted%5Blabel%3D%22clear%22%5D%20%2F%2F%20Not%20in%20Atlas%20nor%20in%20the%20spec%20anymore%0A%7D):

graphviz (4)

### All Submissions:

* [X] Have you signed our [CLA](https://www.mongodb.com/legal/contributor-agreement)?

### CODE REVIEW

File: pkg/controller/atlasdeployment/searchindex_test.go [hunk 1]
```diff
@@ -697,8 +697,50 @@ func Test_searchIndexReconciler(t *testing.T) {
 			ID:                         pointer.MakePtr("testID"),
 			Status:                     nil,
 		}
-		result := reconciler.reconcileInternal(idxInAKO, idxInAtlas)
+		result := reconciler.reconcileInternal("", idxInAKO, idxInAtlas)
 		assert.False(t, result.IsOk())
 		assert.True(t, reconciler.ctx.HasReason(status.SearchIndexStatusError))
 	})
+
+	t.Run("DROP: Must clear if the index disappeared from Atlas", func(t *testing.T) {
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/searchindex_test.go:705]: nit: don't use uppercase

File: pkg/controller/atlasdeployment/searchindex_test.go [hunk 2]
```diff
@@ -697,8 +697,50 @@ func Test_searchIndexReconciler(t *testing.T) {
...
+				Context: context.Background(),
+			},
+			deployment:    testCluster,
+			k8sClient:     nil,
+			projectID:     "",
+			indexName:     "testIndexName",
+			searchService: atlasSearch,
+		}
+		// In previous status, but not found in Atlas now
+		idxInAtlas := &searchindex.SearchIndex{
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/searchindex_test.go:736]: this test is missing the case where `idxInAtlas` is empty which can happen.

File: pkg/controller/atlasdeployment/searchindex.go [hunk 3]
```diff
@@ -77,15 +76,15 @@ func (sr *searchIndexReconciler) Reconcile(spec *akov2.SearchIndex, previous *st
 		}
 	}
 
-	return sr.reconcileInternal(stateInAKO, stateInAtlas)
+	return sr.reconcileInternal(name, stateInAKO, stateInAtlas)
 }
 
-func (sr *searchIndexReconciler) reconcileInternal(stateInAKO, stateInAtlas *searchindex.SearchIndex) workflow.Result {
+func (sr *searchIndexReconciler) reconcileInternal(idxName string, stateInAKO, stateInAtlas *searchindex.SearchIndex) workflow.Result {
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/searchindex.go:82]: suggestion: `idxName` implies many things. The `idx` prefix most likely stands for "index" but is not obvious to the reader. Let's spell this out to `searchIndexName` which makes it very explicit.

File: pkg/controller/atlasdeployment/searchindex.go [hunk 4]
```diff
@@ -28,20 +28,19 @@ type searchIndexReconciler struct {
 
 func (sr *searchIndexReconciler) Reconcile(spec *akov2.SearchIndex, previous *status.DeploymentSearchIndexStatus) workflow.Result {
 	var stateInAtlas, stateInAKO *searchindex.SearchIndex
+	name := ""
```
Comment (s-urbaniak) [inline pkg/controller/atlasdeployment/searchindex.go:31]: side note: if `previous == nil` then `reconcileInternal` will be invoked with empty string `""`. This needs to be reflected in the tests and asserted how the system should behave.

File: pkg/controller/atlasdeployment/searchindex_test.go (hunk 2 above)
Comment (josvazg) [reply pkg/controller/atlasdeployment/searchindex_test.go:736]: tested

**Review by s-urbaniak — APPROVED** [review]

File: pkg/controller/atlasdeployment/searchindex.go [hunk 5]
```diff
@@ -77,15 +76,15 @@ func (sr *searchIndexReconciler) Reconcile(spec *akov2.SearchIndex, previous *st
...
 
-func (sr *searchIndexReconciler) reconcileInternal(stateInAKO, stateInAtlas *searchindex.SearchIndex) workflow.Result {
+func (sr *searchIndexReconciler) reconcileInternal(indexName string, stateInAKO, stateInAtlas *searchindex.SearchIndex) workflow.Result {
 	sr.ctx.Log.Debugf("starting reconciliation for index '%s'", sr.indexName)
 	defer sr.ctx.Log.Debugf("finished reconciliation for index '%s'", sr.indexName)
 
-	emptyInAtlas := stateInAtlas == nil
-	emptyInAKO := stateInAKO == nil
+	inAtlas := stateInAtlas != nil
+	inSpec := stateInAKO != nil
```
Comment (igor-karpukhin) [inline pkg/controller/atlasdeployment/searchindex.go:88]: Why? But it's okay, I guess

**Review by igor-karpukhin — APPROVED** [review]

File: pkg/controller/atlasdeployment/searchindex.go (hunk 5 above)
Comment (josvazg) [reply pkg/controller/atlasdeployment/searchindex.go:88]: I'd say Spec is more accurate and avoids explaining an acronym. We use AKO too much. But I agree we are not consistent, even in this PR.

**Pull request merged**